pkg/operator/environmentchecker/environmentcheckcontroller.go

```diff
@@ -0,0 +1,409 @@
+}
+
+func (c *EnvironmentCheckController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	klog.V(4).Infof("%s: sync started", c.name)
+	defer klog.V(4).Infof("%s: sync complete", c.name)
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	if opSpec.ManagementState != operatorapi.Managed {
+		klog.V(4).Infof("%s: ManagementState is %s, skipping", c.name, opSpec.ManagementState)
+		return nil
+	}
+
+	infra, err := c.infraLister.Get(infraGlobalName)
+	if err != nil {
+		return err
+	}
+
+	if infra.Status.PlatformStatus == nil {
+		klog.V(4).Infof("Unknown platform: infrastructure status.platformStatus is nil")
+		return nil
+	}
+
+	if infra.Status.PlatformStatus.Type != ocpv1.VSpherePlatformType {
+		klog.V(4).Infof("Unsupported platform: infrastructure status.platformStatus.type is %s", infra.Status.PlatformStatus.Type)
+		return nil
+	}
+
+	delay, result, checkRan, ocpError := c.runClusterCheck(ctx, infra)
+	if ocpError != nil {
+		return ocpError
+	}
+
+	// if checks did not run
+	if !checkRan {
+		return nil
+	}
+	queue := syncContext.Queue()
+	queueKey := syncContext.QueueKey()
+
+	klog.V(2).Infof("Scheduled the next check in %s", delay)
+	time.AfterFunc(delay, func() {
+		queue.Add(queueKey)
+	})
+
+	if c.shouldDegradeCluster(result) {
+		return result.CheckError
+	}
+
+	// if operand was not started previously and block upgrade is false and clusterdegrade is also false
+	// then and only then we should start CSI driver operator
+	if !c.operandControllerStarted && !result.BlockUpgrade && !result.ClusterDegrade {
+		go c.runConditionalController(ctx)
+		c.operandControllerStarted = true
+	}
+
+	// sync storage class
+	storageClassSyncResult := c.storageClassController.Sync(ctx, c.vSphereConnection)
+
+	if c.shouldDegradeCluster(storageClassSyncResult) {
+		return storageClassSyncResult.CheckError
+	}
+
+	storageClassUpdateConditionError := c.updateConditions(storageClassControllerName, storageClassSyncResult)
+
+	if storageClassUpdateConditionError != nil {
+		return storageClassUpdateConditionError
+	}
+
+	return c.updateConditions(c.name, result)
+}
+
+func (c *EnvironmentCheckController) shouldDegradeCluster(result checks.ClusterCheckResult) bool {
+	if result.ClusterDegrade {
+		return true
+	}
+
+	if result.BlockUpgrade {
+		// if we were supposed to only block upgrades but for some reason we have created CSIDriver and storageclass
+		// in past and future syncing fails then we should degrade the cluster
+		csiDriver, err := c.csiDriverLister.Get(utils.VSphereDriverName)
+
+		// if we can't fetch CSI driver lets not degrade the cluster, if cluster indeed needs degrading we will catch in
+		// next sync
+		if err != nil {
+			return false
+		}
+		annotations := csiDriver.GetAnnotations()
+		if _, ok := annotations[utils.OpenshiftCSIDriverAnnotationKey]; ok {
+			return true
+		}
+		sc, err := c.scLister.Get(storageClassName)
```

Comment:
I don't know if we should annotate storageclass we create, so as we can guarantee that it is OCP that created the storageclass. @bertinatto @jsafrane